Add XLA Profiler section to docs (#11436)
PyTorchLightning/pytorch-lightning: docs/source/advanced/profiler.rst

Or::

    python -c 'import torch; print(torch.autograd.profiler.load_nvprof("trace_name.prof"))'

XLA Profiling
-------------

:class:`~pytorch_lightning.profiler.xla.XLAProfiler` will help you debug and optimize training

> workload performance for your models using Cloud TPU performance tools.

Reviewer: ```suggestion
workload performance for your models using Cloud TPU performance tools.

.. code-block:: python
    # by passing a string
    trainer = Trainer(..., profiler="xla")

    # or by passing an instance
    from pytorch_lightning.profiler.xla import XLAProfiler

    profiler = XLAProfiler(port=?)
    trainer = Trainer(..., profiler=profiler)
```

what is port in `XLAProfiler`?

Author: Updated! Added it to docstring

> workload performance for your models using Cloud TPU performance tools.

Author: We mean it in the generic manner here. The profiling would work the same for all running stage processes. As it's focus is mainly towards performance summary and finding bottlenecks in the XLA operations.

Manual capture via TensorBoard
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The following instructions are for capturing trace from a running program:
> 
> 0. This `guide <https://cloud.google.com/tpu/docs/pytorch-xla-performance-profiling-tpu-vm#tpu-vm>`_ will
> help you with the Cloud TPU setup with the required installations
> 
> 1. Start a `TensorBoard <https://www.tensorflow.org/tensorboard>`_ server. You could view the TensorBoard output at ``http://localhost:9001`` on your local machine, and then open the
> ``PROFILE`` plugin from the top right dropdown or open http://localhost:9001/#profile
> 
> .. code-block:: bash
> 
>     tensorboard --logdir ./tensorboard --port 9001
> 
> 2. Once the code you'd like to profile is running, click on ``CAPTURE PROFILE`` button. You could enter

Reviewer: ```suggestion
The following instructions are for capturing traces from a running program:

0. This `guide <https://cloud.google.com/tpu/docs/pytorch-xla-performance-profiling-tpu-vm#tpu-vm>`_ will
help you with the Cloud TPU setup with the required installations

1. Start a `TensorBoard <https://www.tensorflow.org/tensorboard>`_ server. You could view the TensorBoard output at ``http://localhost:9001`` on your local machine, and then open the
``PROFILE`` plugin from the top right dropdown or open http://localhost:9001/#profile

.. code-block:: bash

    tensorboard --logdir ./tensorboard --port 9001

2. Once the code you'd like to profile is running, click on the ``CAPTURE PROFILE`` button. You could enter
```
Grammarly 🙈

Or::

    python -c 'import torch; print(torch.autograd.profiler.load_nvprof("trace_name.prof"))'

XLA Profiling
-------------

> `XLA Profiler <https://pytorch-lightning.readthedocs.io/en/latest/api/pytorch_lightning.profiler.XLAProfiler.html#pytorch_lightning.profiler.XLAProfiler>`_ will help you debug and optimize training

Reviewer: ```suggestion
:class:`~pytorch_lightning.profiler.xla.XLAProfiler` will help you debug and optimize training
```

workload performance for your models using Cloud TPU performance tools.

> ##############################
> Manual capture via TensorBoard
> ##############################

Reviewer: ```suggestion
Manual capture via TensorBoard
^^^^^^^^^^^^^^^^^^^^^^^^
```

The following instructions are for capturing trace from a running program:

0. This `guide <https://cloud.google.com/tpu/docs/pytorch-xla-performance-profiling-tpu-vm#tpu-vm>`_ will
help you with the Cloud TPU setup with the required installations

> 1. Start a TensorBoard Server. You could view the TensorBoard output at http://localhost:9001 on your local machine, and then open the

Reviewer: ```suggestion
1. Start a `TensorBoard <https://www.tensorflow.org/tensorboard>`_ server. You could view the TensorBoard output at ``http://localhost:9001`` on your local machine, and then open the
```

Or::

    python -c 'import torch; print(torch.autograd.profiler.load_nvprof("trace_name.prof"))'

XLA Profiling
-------------

:class:`~pytorch_lightning.profiler.xla.XLAProfiler` will help you debug and optimize training
workload performance for your models using Cloud TPU performance tools.

.. code-block:: python

    # by passing the `XLAProfiler` alias
    trainer = Trainer(..., profiler="xla")

    # or by passing an instance
    from pytorch_lightning.profiler import XLAProfiler

    profiler = XLAProfiler(port=9012)
    trainer = Trainer(..., profiler=profiler)

Manual capture via TensorBoard

> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer 2: @rohitgr7 started with standardizing the titles, let's make it capitalized :)) sorry for the nit ^^

The following instructions are for capturing traces from a running program:

0. This `guide <https://cloud.google.com/tpu/docs/pytorch-xla-performance-profiling-tpu-vm#tpu-vm>`_ will

> help you with the Cloud TPU setup with the required installations

Reviewer 2: ```suggestion
help you with the Cloud TPU setup with the required installations.
```

> 1. Start a `TensorBoard <https://www.tensorflow.org/tensorboard>`_ server. You could view the TensorBoard output at ``http://localhost:9001`` on your local machine, and then open the

Reviewer 3: Can we add a gif to show the steps?

> ``PROFILE`` plugin from the top right dropdown or open http://localhost:9001/#profile

Reviewer 2: the link could escaped/formatted

.. code-block:: bash

    tensorboard --logdir ./tensorboard --port 9001

> 2. Once the code you'd like to profile is running, click on the ``CAPTURE PROFILE`` button. You could enter

Reviewer 2: ```suggestion
2. Once the code you'd like to profile is running, click on the ``CAPTURE PROFILE`` button. Enter
```

> ``localhost:9012`` (default port for XLA Profiler) as the Profile Service URL. Then, enter

Reviewer 3: ```suggestion
``localhost:9001`` (default port for XLA Profiler) as the Profile Service URL. Then, enter
```

Reviewer 3: not sure ?

Reviewer 4: https://github.com/PyTorchLightning/pytorch-lightning/pull/11436#discussion_r783869996

> ``localhost:9012`` (default port for XLA Profiler) as the Profile Service URL. Then, you could enter

Reviewer: quick ques: are these ports used differently?

Reviewer 4: As I read the Google Cloud blog post, it should be the same, i.e., 9001. https://cloud.google.com/tpu/docs/pytorch-xla-performance-profiling-tpu-vm#tpu-vm

> ``localhost:9012`` (default port for XLA Profiler) as the Profile Service URL. Then, you could enter

Reviewer 2: ```suggestion
``localhost:9012`` (default port for XLA Profiler) as the Profile Service URL. Then, enter
```

the number of milliseconds for the profiling duration, and click ``CAPTURE``

> 3. Make sure the code is running, while you are trying to capture the traces. Also, it would lead to better
> performance insights if the profiling duration is longer than the step time

Reviewer 2: ```suggestion
3. Make sure the code is running while you are trying to capture the traces. Also, it would lead to better
performance insights if the profiling duration is longer than the step time.
```

> 4. Once the capture is finished, the page will refresh and you could browse through the insights using the
> ``Tools`` dropdown at the top left

Reviewer 2: ```suggestion
4. Once the capture is finished, the page will refresh and you can browse through the insights using the
``Tools`` dropdown at the top left
```

:class:`~pytorch_lightning.profiler.xla.XLAProfiler` will help you debug and optimize training
workload performance for your models using Cloud TPU performance tools.

.. code-block:: python

    # by passing the `XLAProfiler` alias
    trainer = Trainer(..., profiler="xla")

    # or by passing an instance

>     from pytorch_lightning.profiler.xla import XLAProfiler

Reviewer 5: should we make the profiler imports all be exposed at the lightning.profiler module? this would be consistent with what we do for other components. fyi @daniellepintz

Author: Yup, updated! Thanks

    python -c 'import torch; print(torch.autograd.profiler.load_nvprof("trace_name.prof"))'

XLA Profiler
============

:class:`~pytorch_lightning.profiler.xla.XLAProfiler` will help you debug and optimize training
workload performance for your models using Cloud TPU performance tools.

.. code-block:: python

    # by passing the `XLAProfiler` alias
    trainer = Trainer(..., profiler="xla")

    # or by passing an instance
    from pytorch_lightning.profiler import XLAProfiler

>     profiler = XLAProfiler(port=9001)

Reviewer 4: @kaushikb11 I updated the port from 9012 to 9001 here in 37b1601. Correct? If not, please revert the commit...

Author: It's right!
